pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -174,27 +209,93 @@ func (c *Controller) syncGKENetworkParamSet(ctx context.Context, key string) err
 	}
 
 	params := obj.(*networkv1.GKENetworkParamSet)
+	originalParams := params.DeepCopy()
 
-	// TODO: Enable finalizer addition when finalizer deletion is added.
-	// addFinalizerToGKENetworkParamSet(params)
-	// update will be done once in deferred function call
-	// if err := c.updateGKENetworkParamSet(ctx, params); err != nil {
-	// 	return err
-	// }
+	defer func() {
+		if !reflect.DeepEqual(originalParams, params) {
+			if updateErr := c.updateGKENetworkParamSet(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+
+		if !reflect.DeepEqual(originalParams.Status, params.Status) {
+			if updateErr := c.updateGKENetworkParamSetStatus(ctx, params); updateErr != nil {
+				err = multierror.Append(updateErr, err)
+			}
+		}
+	}()
+
+	if params.DeletionTimestamp != nil {
+		// GKENetworkParamSet is being deleted, handle the delete event
+		return c.handleGKENetworkParamSetDelete(ctx, params)
+	}
 
-	subnet, err := c.gceCloud.GetSubnetwork(c.gceCloud.Region(), params.Spec.VPCSubnet)
+	addFinalizerInPlace(params)
+
+	subnet, subnetValidation := c.getAndValidateSubnet(ctx, params)
+	params.Status.Conditions = updateConditions(params.Status.Conditions, subnetValidation.toCondition())
+	if !subnetValidation.IsValid {
+		return
```

Comment:
```suggestion
		return fmt.Errorf("subnet validation failed blah blah")
```

Reply:
or return nil and log the message

Reply:
The result of these validations is always put into the conditions of the object which is not a program error. If there was an error while updating the object's status which contains the condition, that would be an error worth logging (and is already done so).